two-factor.php

```diff
@@ -52,67 +27,88 @@ function wpcom_vip_force_two_factor() {
 		return false;
 	}
 
-	return apply_filters( 'wpcom_vip_force_two_factor', false );
+	return apply_filters( 'wpcom_vip_is_two_factor_forced', false );
 }
 
 function wpcom_vip_enforce_two_factor_plugin() {
 	if ( is_user_logged_in() ) {
+		$limited = current_user_can( 'edit_posts' );
+
+		if ( VIP_IS_AFTER_2FA_TIME_GATE ) {
+			// Calculate current_user_can outside map_meta_cap to avoid callback loop
+			add_filter( 'wpcom_vip_is_two_factor_forced', function() use ( $limited ) {
+				return $limited;
+			} );
+		} else if ( $limited && ! Two_Factor_Core::is_user_using_two_factor() ) {
+			add_action( 'admin_notices', 'wpcom_vip_two_factor_prep_admin_notice' );
+		}
+
 		add_action( 'admin_notices', 'wpcom_vip_two_factor_admin_notice' );
-		add_filter( 'map_meta_cap', 'wpcom_vip_two_factor_filter_caps' );
+		add_filter( 'map_meta_cap', 'wpcom_vip_two_factor_filter_caps', 0, 4 );
 	}
 }
 
+add_action( 'muplugins_loaded', 'wpcom_enable_two_factor_plugin' );
 function wpcom_enable_two_factor_plugin() {
-	if ( wpcom_vip_is_jetpack_sso_enabled() ) {
-		if ( ! class_exists( 'Jetpack_Force_2FA' ) ) {
-			wpcom_vip_load_plugin( 'jetpack-force-2fa' );
-		}
-
-		// Prevent lockout if we end up with both jetpack-2fa and twofactor enabled at the same time.
-		if ( class_exists( 'Two_Factor_Core' ) ) {
-			remove_action( 'wp_login', array( 'Two_Factor_Core', 'wp_login' ) );
-		}
-	} else {
-		wpcom_vip_load_plugin( 'two-factor' );
-		add_action( 'set_current_user', 'wpcom_vip_enforce_two_factor_plugin' );
-	}
+	wpcom_vip_load_plugin( 'two-factor' );
+	add_action( 'set_current_user', 'wpcom_vip_enforce_two_factor_plugin' );
 }
-// TODO: Enable this :)
-//add_action( 'muplugins_loaded', 'wpcom_enable_two_factor_plugin' );
 
 /**
  * Filter Caps
  *
  * Remove caps for users without two-factor enabled so they are treated as a Contributor.
  */
-function wpcom_vip_two_factor_filter_caps( $caps ) {
-	if ( wpcom_vip_force_two_factor() ) {
-		// Use a hard-coded list of caps that closely match a Contributor role.
-		// The hard-coded list avoids issues if the Contributor role doesn't exist or is modified.
-		$contributor_caps = [
-			'edit_posts' => true,
-			'read' => true,
-			'level_1' => true,
-			'level_0' => true,
+function wpcom_vip_two_factor_filter_caps( $caps, $cap, $user_id, $args ) {
+	if ( wpcom_vip_is_two_factor_forced() ) {
+		// Use a hard-coded list of caps that give just enough access to set up 2FA
+		$subscriber_caps = [
+			'read',
```

Comment:
Do we need `level_0` in there just-in-case?

```suggestion
			'read',
			'level_0',
```

https://codex.wordpress.org/Roles_and_Capabilities#User_Levels